src/nnet3/nnet-computation.cc

```diff
@@ -414,6 +648,209 @@ void NnetComputation::Print(std::ostream &os, const Nnet &nnet) const {
+  }
+
+  size_t num_submatrices;
+  ExpectToken(is, binary, "<NumSubMatrices>");
+  ReadBasicType(is, binary, &num_submatrices);
+  KALDI_ASSERT(num_submatrices >= 0);
+  submatrices.resize(num_submatrices);
+  ExpectToken(is, binary, "<SubMatrices>");
+  for (size_t c = 0; c < num_submatrices; c++) {
+    submatrices[c].Read(is, binary);
+  }
+
+
+  size_t num_component_precomputed_indexes;
+  ExpectToken(is, binary, "<NumComponentPrecomputedIndexes>");
+  ReadBasicType(is, binary, &num_component_precomputed_indexes);
+  KALDI_ASSERT(num_component_precomputed_indexes >= 0);
+  component_precomputed_indexes.resize(num_component_precomputed_indexes);
+  ExpectToken(is, binary, "<ComponentPrecomputedIndexes>");
+  std::vector<ComponentPrecomputedIndexes*> component_precomputed_indexes_tmp;
+  for (size_t c = 0; c < num_component_precomputed_indexes; c++) {
+    bool is_null; // a boolean indicating whether the pointer should be NULL.
+    ReadBasicType(is, binary, &is_null); 
+    if (!is_null) {
+      ComponentPrecomputedIndexes* p = ComponentPrecomputedIndexes::ReadNew(is, binary);
+      component_precomputed_indexes_tmp.push_back(p);
+    } else {
+      component_precomputed_indexes_tmp.push_back(NULL);
+    }
+  }
+  component_precomputed_indexes = component_precomputed_indexes_tmp;
+
+  size_t num_indexes;
+  ExpectToken(is, binary, "<NumIndexes>");
+  ReadBasicType(is, binary, &num_indexes);
+  KALDI_ASSERT(num_indexes >= 0);
+  indexes.resize(num_indexes);
+  ExpectToken(is, binary, "<Indexes>");
+  for (size_t c = 0; c < num_indexes; c++) {
+    ReadIntegerVector(is, binary, &(indexes[c]));
+  }
+
+  size_t num_indexes_multi;
+  ExpectToken(is, binary, "<NumIndexesMulti>");
+  ReadBasicType(is, binary, &num_indexes_multi);
+  KALDI_ASSERT(num_indexes_multi >= 0);
+  indexes_multi.resize(num_indexes_multi);
+  ExpectToken(is, binary, "<IndexesMulti>");
+  for (size_t c = 0; c < num_indexes_multi; c++) {
+    ReadIntegerPairVector(is, binary, &(indexes_multi[c]));
+  }
+
+  size_t num_indexes_ranges;
+  ExpectToken(is, binary, "<NumIndexesRanges>");
+  ReadBasicType(is, binary, &num_indexes_ranges);
+  KALDI_ASSERT(num_indexes_ranges >= 0);
+  indexes_ranges.resize(num_indexes_ranges);
+  ExpectToken(is, binary, "<IndexesRanges>");
+  for (size_t c = 0; c < num_indexes_ranges; c++) {
+    ReadIntegerPairVector(is, binary, &(indexes_ranges[c]));
+  }
+
+  size_t num_input_output_info;
+  ExpectToken(is, binary, "<NumInputOutputInfo>");
+  ReadBasicType(is, binary, &num_input_output_info);
+  KALDI_ASSERT(num_input_output_info >= 0);
+  input_output_info.clear();
+  ExpectToken(is, binary, "<InputOutputInfo>");
+  for (size_t c = 0; c < num_input_output_info; c++) {
+    int32 key;
+    std::pair<int32, int32> val;
+    ReadBasicType(is, binary, &key);
+    ReadBasicType(is, binary, &(val.first));
+    ReadBasicType(is, binary, &(val.second));
+    input_output_info.insert(std::make_pair<int32, std::pair<int32, int32> >(key, val));
```

Comment:
Thanks. Have committed a fix. Dan, can you remind me what else issues does make_pair have? I'll avoid using it in the future.

On Mar 5, 2016, at 9:44 PM, Daniel Povey <notifications@github.com> wrote:

> In src/nnet3/nnet-computation.cc:
> 
> > +    ReadIntegerPairVector(is, binary, &(indexes_ranges[c]));
> > +  }
> > +
> > +  size_t num_input_output_info;
> > +  ExpectToken(is, binary, "<NumInputOutputInfo>");
> > +  ReadBasicType(is, binary, &num_input_output_info);
> > +  KALDI_ASSERT(num_input_output_info >= 0);
> > +  input_output_info.clear();
> > +  ExpectToken(is, binary, "<InputOutputInfo>");
> > +  for (size_t c = 0; c < num_input_output_info; c++) {
> > +    int32 key;
> > +    std::pair<int32, int32> val;
> > +    ReadBasicType(is, binary, &key);
> > +    ReadBasicType(is, binary, &(val.first));
> > +    ReadBasicType(is, binary, &(val.second));
> > +    input_output_info.insert(std::make_pair<int32, std::pair<int32, int32> >(key, val));
> I think there are other issues with make_pair, we've tried to avoid it. Samuel, you should be using the following, which invokes the constructor: std::pair<int32, std::pair<int32, int32> >(key, val)
> …
> —
> Reply to this email directly or view it on GitHub.
>